core/src/main/java/org/apache/calcite/rex/RexInterpreter.java

```diff
@@ -213,7 +213,7 @@ private Comparable extract(RexCall call, List<Comparable> values) {
     final int v2;
     if (v instanceof Long) {
       // TIMESTAMP
-      v2 = (int) (((Long) v) / TimeUnit.DAY.multiplier.longValue());
+      v2 = Math.toIntExact(((Long) v) / TimeUnit.DAY.multiplier.longValue());
```

Comment:
One think I don't understand here is `v2` is still used as `long` in the following code `DateTimeUtils.unixDateExtract(timeUnitRange, v2);`. Maybe the right fix is do not do long -> int conversion?

Reply:
> One think I don't understand here is `v2` is still used as `long` in the following code `DateTimeUtils.unixDateExtract(timeUnitRange, v2);`. Maybe the right fix is do not do long -> int conversion?

`v2 = (Integer) v; ` should we cast to `long` too? If it was you would have to handle the `int` and `long` cases separately for `DateTimeUtils.unixDateExtract (timeUnitRange, v2)`;

Reply:
Why do you think a integer days would overflow ? Can you give a test case ?

Reply:
Based on math, if you use max_long / milliseoncd_per_day, the number will still exceed int_max. So technically overflow exists.

Reply:
What is the upper/max value of timestamp million-seconds ?